packages/eslint-plugin-baseui/src/deprecated-theme-api.js

```diff
@@ -0,0 +1,434 @@
+          let fix = null;
+          let messageId = MESSAGES.deprecateThemeProperty.id;
+
+          // If there is a possible replacement for this deprecated property,
+          // add a fix to replace with a new semantic property.
+          // TODO: use `suggest` instead, currently not supported in VSCode extension:
+          // https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions
+          // https://github.com/microsoft/vscode-eslint/issues/806
+          if (deprecatedThemeProperties[identifier].replacement) {
+            messageId = MESSAGES.replaceThemeProperty.id;
+            data.new = deprecatedThemeProperties[identifier].replacement;
+            fix = function(fixer) {
+              return fixer.replaceText(
+                node,
+                deprecatedThemeProperties[identifier].replacement,
+              );
+            };
+          }
+
+          // Check if we are setting a property in createTheme:overrides
+          // Ex: createTheme({}, {datepickerBackground: "pink"});
+          if (
+            context
+              .getAncestors()
+              .some(
+                node =>
+                  node.type === 'CallExpression' &&
+                  node.callee.name === 'createTheme',
+              ) &&
+            node.parent.type === 'Property' &&
+            node === node.parent.key
+          ) {
+            // A deprecated property is being set in createTheme.
+            // Assuming a violation.
+            context.report({
+              node,
+              messageId,
+              data,
+              fix,
+            });
+            return;
+          }
+
+          // Exit early if we are not in a function.
+          // We don't have access to the theme outside of functions.
+          // Ex: styled, withStyle, overrides, function component (useStyletron)
+          if (scope.type !== 'function') {
+            return;
+          }
+
+          // Exit early if the identifier is a key in an object expression
+          // Ex: { background: $theme.colors.white }
+          if (
+            node.parent.type === 'Property' &&
+            node.parent.parent.type === 'ObjectExpression' &&
+            node === node.parent.key
+          ) {
+            return;
+          }
+
+          // Exit early if identifier is a value in an object destructuing pattern
+          // Ex: ({$theme: {colors: {background}}}) => {}
+          //  - background will match twice, as both a key and as a value, so we skip the value match
+          if (
+            node.parent.type === 'Property' &&
+            node.parent.parent.type === 'ObjectPattern' &&
+            node === node.parent.value
+          ) {
+            return;
+          }
+
+          // Check if our identifier is accessing properties on a theme, $theme, or its theme concern.
+          // Ex: theme.colors.foreground
+          // Ex: props.$theme.colors.foreground
+          // Ex: colors.foreground
```

Comment:
Is it possible to find all `styled`, `withStyle`, and `useStyletron` instances.. then identify if `($)theme` is being used as a function argument or is the second array value returned from `useStyletron`? I get a bit nervous about the 'name' heuristic and how this has the potential to raise false positives. Does this restrict any object named `theme` or `colors` unrelated to baseui? I would much prefer to be more passive in matching so that developers do not turn the feature off or become hostile towards it

Reply:
After refactor it works close to this - still keys off the deprecated property as an identifier but then checks scope to verify it is inside of a styled|withStyle|overrides style function or that useStyletron is used.